subprojects/launcher/src/integTest/groovy/org/gradle/launcher/continuous/ContinuousBuildGateIntegrationTest.groovy

```diff
@@ -0,0 +1,130 @@
+                    this.gateKeeper = buildGate.createGateKeeper()
+                }
+                void run() {
+                    boolean stop = false
+                    while (!stop) {
+                        def command = "${server.uri("command")}".toURL().text
+                        if (command == "open") {
+                            println("[GK] Open gate")
+                            gateKeeper.open()
+                        } else if (command == "close") {
+                            println("[GK] Close gate")
+                            gateKeeper.close()
+                        } else if (command == "stop") {
+                            println("[GK] stop")
+                            stop = true
+                        } else {
+                            println("[GK] ? " + command)
+                        }
+                    }
+                }
+            }
+            class BuildGateKeeperStarter {
+                static BuildGateKeeper gateKeeper
+                static void start(BuildGateToken buildGate) {
+                    if (gateKeeper == null) {
+                        println "Starting gatekeeper"
+                        gateKeeper = new BuildGateKeeper(buildGate)
+                        new Thread(gateKeeper).start()
+                    }
+                }
+            }
+
+            def buildGate = gradle.services.get(BuildGateToken)
+            BuildGateKeeperStarter.start(buildGate)
+
+            class SimpleTask extends DefaultTask {
+                @InputFile
+                File inputFile = project.file("input.txt")
+                
+                @OutputFile
+                File outputFile = new File(project.buildDir, "output.txt")
+                
+                @TaskAction
+                void generate() {
+                    outputFile.text = inputFile.text
+                } 
+            }
+            
+            def pendingChangesManager = gradle.services.get(${PendingChangesManager.canonicalName})
+            pendingChangesManager.addListener new ${SingleFirePendingChangesListener.canonicalName}({
+                ${server.callFromBuild("pending")}
+            } as ${PendingChangesListener.canonicalName})
+            
+            task work(type: SimpleTask)
+        """
+
+        file("input.txt").text = "start"
+    }
+
+    def "build only starts when gate is opened"() {
+        server.expect(server.resource("command", "close"))
+        def command = server.expectAndBlock(server.resource("command", "open"))
+
+        def inputFile = file("input.txt")
+        def outputFile = file("build/output.txt")
+
+        when:
+        succeeds("work")
+        then:
+        outputFile.text == "start"
+
+        when:
+        // Make some file system changes to inputFile
+        inputFile.text = "changed"
+        then:
+        // Gradle has detected the changes to inputFile
+        def pending = server.expectAndBlock("pending")
```

Comment:
Although unlikely, this is a race condition: the 'pending' call could happen before the `expectAndBlock` is configured.
Instead, it should be:
```
// Set up the block
def pending = server.expectAndBlock('pending')
// Trigger the change that will get us to the block
inputFile.text = 'changed'
// Wait for the block to be reached
pending.waitForAllPendingCalls()
```